In [1]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report
)

In [2]:
heart_df = pd.read_excel(r"D:\DS 8AM\project\heart_disease_cleaned.xlsx")

X = heart_df.drop(columns=["heartStroke"])
y = heart_df["heartStroke"]

In [3]:
#from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

categorical_cols = ["Gender", "Education", "Exercise", "AgeGroup"]

for col in categorical_cols:
    X[col] = le.fit_transform(X[col])

In [4]:
#from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [7]:
X_train.shape

(3396, 17)

In [8]:
X_test.shape

(850, 17)

In [9]:
y_train.value_counts()

heartStroke
0    2872
1     524
Name: count, dtype: int64

In [10]:
y_test.value_counts()

heartStroke
0    727
1    123
Name: count, dtype: int64

### Logistic Regression

In [12]:
# Feature Scaling (Required for Logistic Regression & KNN)

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

# Scale training and testing data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Logistic Regression

lr = LogisticRegression(random_state=42)

lr.fit(X_train_scaled, y_train)

lr_pred = lr.predict(X_test_scaled)

In [13]:
# Evaluate Logistic Regression

print("Accuracy :", round(accuracy_score(y_test, lr_pred), 4))
print("Precision:", round(precision_score(y_test, lr_pred), 4))
print("Recall   :", round(recall_score(y_test, lr_pred), 4))
print("F1 Score :", round(f1_score(y_test, lr_pred), 4))

print("\nConfusion Matrix")
print(confusion_matrix(y_test, lr_pred))

print("\nClassification Report")
print(classification_report(y_test, lr_pred))

Accuracy : 0.8576
Precision: 0.6
Recall   : 0.0488
F1 Score : 0.0902

Confusion Matrix
[[723   4]
 [117   6]]

Classification Report
              precision    recall  f1-score   support

           0       0.86      0.99      0.92       727
           1       0.60      0.05      0.09       123

    accuracy                           0.86       850
   macro avg       0.73      0.52      0.51       850
weighted avg       0.82      0.86      0.80       850



#### Conclusion
Logistic Regression achieved an accuracy of 85.76%, but its recall was only 4.88%, indicating that it failed to identify most patients with heart stroke. Although the overall accuracy appears high due to the imbalanced dataset, the model is not suitable as the final prediction model because it misses a large proportion of positive cases.

### Decision Tree Classifier

In [15]:
dt = DecisionTreeClassifier(random_state=42)

dt.fit(X_train, y_train)

dt_pred = dt.predict(X_test)

In [16]:
# Evaluate Decision Tree

print("Accuracy :", round(accuracy_score(y_test, dt_pred), 4))
print("Precision:", round(precision_score(y_test, dt_pred), 4))
print("Recall   :", round(recall_score(y_test, dt_pred), 4))
print("F1 Score :", round(f1_score(y_test, dt_pred), 4))

print("\nConfusion Matrix")
print(confusion_matrix(y_test, dt_pred))

print("\nClassification Report")
print(classification_report(y_test, dt_pred))

Accuracy : 0.7776
Precision: 0.2537
Recall   : 0.2764
F1 Score : 0.2646

Confusion Matrix
[[627 100]
 [ 89  34]]

Classification Report
              precision    recall  f1-score   support

           0       0.88      0.86      0.87       727
           1       0.25      0.28      0.26       123

    accuracy                           0.78       850
   macro avg       0.56      0.57      0.57       850
weighted avg       0.79      0.78      0.78       850



#### Conclusion
The Decision Tree model achieved a lower accuracy (77.76%) than Logistic Regression (85.76%), but substantially improved recall (27.64% vs. 4.88%). This indicates that the model identifies considerably more patients with heart stroke, making it more suitable for medical prediction despite a reduction in overall accuracy.

### Random Forest Classifier

In [17]:
# Train Random Forest

rf = RandomForestClassifier(
    random_state=42
)

rf.fit(X_train, y_train)

rf_pred = rf.predict(X_test)

In [18]:
# Evaluate Random Forest

print("Accuracy :", round(accuracy_score(y_test, rf_pred), 4))
print("Precision:", round(precision_score(y_test, rf_pred), 4))
print("Recall   :", round(recall_score(y_test, rf_pred), 4))
print("F1 Score :", round(f1_score(y_test, rf_pred), 4))

print("\nConfusion Matrix")
print(confusion_matrix(y_test, rf_pred))

print("\nClassification Report")
print(classification_report(y_test, rf_pred))

Accuracy : 0.8529
Precision: 0.45
Recall   : 0.0732
F1 Score : 0.1259

Confusion Matrix
[[716  11]
 [114   9]]

Classification Report
              precision    recall  f1-score   support

           0       0.86      0.98      0.92       727
           1       0.45      0.07      0.13       123

    accuracy                           0.85       850
   macro avg       0.66      0.53      0.52       850
weighted avg       0.80      0.85      0.80       850



### Conclusions
The Random Forest model has:

    High accuracy (85.29%), similar to Logistic Regression.
    Better recall than Logistic Regression, but still very low.
    Worse recall than Decision Tree.

The confusion matrix shows that it correctly identified only 9 out of 123 patients with heart stroke. This is better than Logistic Regression (6) but much worse than Decision Tree (34).

### K-Nearest Neighbours

In [19]:
# Train KNN

knn = KNeighborsClassifier()

knn.fit(X_train_scaled, y_train)

knn_pred = knn.predict(X_test_scaled)

In [20]:
# Evaluate KNN

print("Accuracy :", round(accuracy_score(y_test, knn_pred), 4))
print("Precision:", round(precision_score(y_test, knn_pred), 4))
print("Recall   :", round(recall_score(y_test, knn_pred), 4))
print("F1 Score :", round(f1_score(y_test, knn_pred), 4))

print("\nConfusion Matrix")
print(confusion_matrix(y_test, knn_pred))

print("\nClassification Report")
print(classification_report(y_test, knn_pred))

Accuracy : 0.8494
Precision: 0.4324
Recall   : 0.1301
F1 Score : 0.2

Confusion Matrix
[[706  21]
 [107  16]]

Classification Report
              precision    recall  f1-score   support

           0       0.87      0.97      0.92       727
           1       0.43      0.13      0.20       123

    accuracy                           0.85       850
   macro avg       0.65      0.55      0.56       850
weighted avg       0.81      0.85      0.81       850



### Model Comparison

In [22]:
# Model Comparison

results = pd.DataFrame({
    "Model": [
        "Logistic Regression",
        "Decision Tree",
        "Random Forest",
        "KNN"
    ],
    "Accuracy": [
        accuracy_score(y_test, lr_pred),
        accuracy_score(y_test, dt_pred),
        accuracy_score(y_test, rf_pred),
        accuracy_score(y_test, knn_pred)
    ],
    "Precision": [
        precision_score(y_test, lr_pred),
        precision_score(y_test, dt_pred),
        precision_score(y_test, rf_pred),
        precision_score(y_test, knn_pred)
    ],
    "Recall": [
        recall_score(y_test, lr_pred),
        recall_score(y_test, dt_pred),
        recall_score(y_test, rf_pred),
        recall_score(y_test, knn_pred)
    ],
    "F1 Score": [
        f1_score(y_test, lr_pred),
        f1_score(y_test, dt_pred),
        f1_score(y_test, rf_pred),
        f1_score(y_test, knn_pred)
    ]
})

results = results.round(4)

results.sort_values(by="F1 Score", ascending=False)

,Model,Accuracy,Precision,Recall,F1 Score
1,Decision Tree,0.7776,0.2537,0.2764,0.2646
3,KNN,0.8494,0.4324,0.1301,0.2000
2,Random Forest,0.8529,0.4500,0.0732,0.1259
0,Logistic Regression,0.8576,0.6000,0.0488,0.0902


# Baseline Model Comparison

Four classification models were trained and evaluated on the heart disease dataset:

- Logistic Regression
- Decision Tree
- Random Forest
- K-Nearest Neighbors (KNN)

The models were evaluated using Accuracy, Precision, Recall, and F1 Score. These metrics provide different perspectives on model performance, especially for imbalanced datasets.

In [23]:
# Baseline Model Comparison

baseline_results = pd.DataFrame({
    "Model": [
        "Logistic Regression",
        "Decision Tree",
        "Random Forest",
        "K-Nearest Neighbors"
    ],
    "Accuracy": [0.8576, 0.7776, 0.8529, 0.8494],
    "Precision": [0.6000, 0.2537, 0.4500, 0.4324],
    "Recall": [0.0488, 0.2764, 0.0732, 0.1301],
    "F1 Score": [0.0902, 0.2646, 0.1259, 0.2000]
})

baseline_results

,Model,Accuracy,Precision,Recall,F1 Score
0,Logistic Regression,0.8576,0.6000,0.0488,0.0902
1,Decision Tree,0.7776,0.2537,0.2764,0.2646
2,Random Forest,0.8529,0.4500,0.0732,0.1259
3,K-Nearest Neighbors,0.8494,0.4324,0.1301,0.2000


## Findings

### Logistic Regression
- Achieved the highest Accuracy (85.76%) and Precision (60.00%).
- However, it identified only 4.88% of actual heart stroke cases, making it unsuitable for medical prediction despite its high overall accuracy.

### Decision Tree
- Achieved the highest Recall (27.64%) and F1 Score (26.46%).
- Although its overall Accuracy (77.76%) was the lowest, it detected the highest number of heart stroke cases among all baseline models.

### Random Forest
- Produced high Accuracy (85.29%), similar to Logistic Regression.
- Recall remained low (7.32%), indicating that most positive cases were still missed.

### K-Nearest Neighbors (KNN)
- Provided a balanced performance with Accuracy of 84.94% and Recall of 13.01%.
- Performed better than Logistic Regression and Random Forest in detecting positive cases but was still inferior to Decision Tree.

### Overall Observation
The dataset is imbalanced, with approximately 85% healthy participants and 15% heart stroke cases. Consequently, Accuracy alone is not a reliable evaluation metric. Recall and F1 Score provide a better assessment of the models' ability to identify patients with heart disease.

Among the four baseline models, the Decision Tree Classifier demonstrated the best overall performance because it achieved the highest Recall and F1 Score, making it the most suitable baseline model for this healthcare classification problem.

In [26]:
# Balanced Logistic Regression

lr_balanced = LogisticRegression(
    random_state=42,
    class_weight="balanced"
)

lr_balanced.fit(X_train_scaled, y_train)

lr_balanced_pred = lr_balanced.predict(X_test_scaled)

In [27]:
# Evaluate Balanced Logistic Regression

print("Accuracy :", round(accuracy_score(y_test, lr_balanced_pred), 4))
print("Precision:", round(precision_score(y_test, lr_balanced_pred), 4))
print("Recall   :", round(recall_score(y_test, lr_balanced_pred), 4))
print("F1 Score :", round(f1_score(y_test, lr_balanced_pred), 4))

print("\nConfusion Matrix")
print(confusion_matrix(y_test, lr_balanced_pred))

print("\nClassification Report")
print(classification_report(y_test, lr_balanced_pred))

Accuracy : 0.6541
Precision: 0.2286
Recall   : 0.5854
F1 Score : 0.3288

Confusion Matrix
[[484 243]
 [ 51  72]]

Classification Report
              precision    recall  f1-score   support

           0       0.90      0.67      0.77       727
           1       0.23      0.59      0.33       123

    accuracy                           0.65       850
   macro avg       0.57      0.63      0.55       850
weighted avg       0.81      0.65      0.70       850



In [28]:
# Balanced Decision Tree

dt_balanced = DecisionTreeClassifier(
    random_state=42,
    class_weight="balanced"
)

dt_balanced.fit(X_train, y_train)

dt_balanced_pred = dt_balanced.predict(X_test)

In [29]:
# Evaluate Balanced Decision Tree

print("Accuracy :", round(accuracy_score(y_test, dt_balanced_pred), 4))
print("Precision:", round(precision_score(y_test, dt_balanced_pred), 4))
print("Recall   :", round(recall_score(y_test, dt_balanced_pred), 4))
print("F1 Score :", round(f1_score(y_test, dt_balanced_pred), 4))

print("\nConfusion Matrix")
print(confusion_matrix(y_test, dt_balanced_pred))

print("\nClassification Report")
print(classification_report(y_test, dt_balanced_pred))

Accuracy : 0.7753
Precision: 0.2639
Recall   : 0.3089
F1 Score : 0.2846

Confusion Matrix
[[621 106]
 [ 85  38]]

Classification Report
              precision    recall  f1-score   support

           0       0.88      0.85      0.87       727
           1       0.26      0.31      0.28       123

    accuracy                           0.78       850
   macro avg       0.57      0.58      0.58       850
weighted avg       0.79      0.78      0.78       850



In [31]:
# Balanced Random Forest

rf_balanced = RandomForestClassifier(
    random_state=42,
    class_weight="balanced"
)

rf_balanced.fit(X_train, y_train)

rf_balanced_pred = rf_balanced.predict(X_test)

In [32]:
# Evaluate Balanced Random Forest

print("Accuracy :", round(accuracy_score(y_test, rf_balanced_pred), 4))
print("Precision:", round(precision_score(y_test, rf_balanced_pred), 4))
print("Recall   :", round(recall_score(y_test, rf_balanced_pred), 4))
print("F1 Score :", round(f1_score(y_test, rf_balanced_pred), 4))

print("\nConfusion Matrix")
print(confusion_matrix(y_test, rf_balanced_pred))

print("\nClassification Report")
print(classification_report(y_test, rf_balanced_pred))

Accuracy : 0.8565
Precision: 0.5714
Recall   : 0.0325
F1 Score : 0.0615

Confusion Matrix
[[724   3]
 [119   4]]

Classification Report
              precision    recall  f1-score   support

           0       0.86      1.00      0.92       727
           1       0.57      0.03      0.06       123

    accuracy                           0.86       850
   macro avg       0.72      0.51      0.49       850
weighted avg       0.82      0.86      0.80       850



# Model Performance Summary

A total of **seven models** were evaluated, including four baseline models and three balanced models. The balanced models were trained using `class_weight='balanced'` to address the class imbalance present in the dataset (approximately 85% healthy participants and 15% heart stroke cases).

| Model | Accuracy | Precision | Recall | F1 Score |
|:------|---------:|----------:|--------:|---------:|
| Logistic Regression | **0.8576** | **0.6000** | 0.0488 | 0.0902 |
| Decision Tree | 0.7776 | 0.2537 | 0.2764 | 0.2646 |
| Random Forest | 0.8529 | 0.4500 | 0.0732 | 0.1259 |
| K-Nearest Neighbors (KNN) | 0.8494 | 0.4324 | 0.1301 | 0.2000 |
| **Balanced Logistic Regression** | 0.6541 | 0.2286 | **0.5854** | **0.3288** |
| Balanced Decision Tree | 0.7753 | 0.2639 | 0.3089 | 0.2846 |
| Balanced Random Forest | 0.8565 | 0.5714 | 0.0325 | 0.0615 |

## Model Comparison

### Logistic Regression
- Achieved the highest baseline **Accuracy (85.76%)** and **Precision (60.00%)**.
- However, it detected only **4.88%** of actual heart stroke cases, making it unsuitable for medical prediction despite its high overall accuracy.

### Decision Tree
- Achieved the highest **Recall (27.64%)** and **F1 Score (26.46%)** among the baseline models.
- It detected more heart stroke cases than the other baseline models, making it the strongest baseline model for this healthcare dataset.

### Random Forest
- Produced high accuracy (**85.29%**) but identified very few positive cases (**Recall = 7.32%**).
- It performed similarly to Logistic Regression but remained weak in detecting heart stroke patients.

### K-Nearest Neighbors (KNN)
- Provided a balanced performance with **84.94% Accuracy** and **13.01% Recall**.
- It outperformed Logistic Regression and Random Forest in detecting positive cases but remained inferior to Decision Tree.

### Balanced Logistic Regression
- Applying `class_weight='balanced'` significantly improved the model.
- Recall increased from **4.88%** to **58.54%**, while the F1 Score increased from **0.0902** to **0.3288**.
- Although overall accuracy decreased to **65.41%**, the model detected the highest number of heart stroke cases, making it the best-performing model for this healthcare classification problem.

### Balanced Decision Tree
- Produced a moderate improvement over the baseline Decision Tree.
- Recall increased from **27.64%** to **30.89%**, while the F1 Score also improved slightly.
- Overall accuracy remained almost unchanged.

### Balanced Random Forest
- Applying class weighting did not improve the model.
- Although accuracy increased slightly, Recall and F1 Score decreased, making it less effective for detecting heart stroke cases than the baseline Random Forest.

## Final Conclusion

The dataset is **imbalanced**, containing substantially more healthy participants than heart stroke cases. Therefore, **Accuracy alone is not an appropriate metric** for selecting the best model. In healthcare applications, **Recall** and **F1 Score** are more important because they measure the model's ability to correctly identify patients with the disease.

Among all evaluated models, **Balanced Logistic Regression** achieved the highest Recall (**58.54%**) and F1 Score (**32.88%**), detecting the greatest number of heart stroke cases. Although it sacrificed some overall accuracy, it provided the best balance between identifying positive cases and maintaining acceptable predictive performance.

Therefore, **Balanced Logistic Regression** was selected as the final model for deployment in the Streamlit application.

# Saving the Final Model

The best-performing model (**Balanced Logistic Regression**) is saved using **Joblib**. Saving the trained model allows it to be reused later in the Streamlit application without retraining.

In [33]:
# Save the Final Model

import joblib

# Save the trained model
joblib.dump(lr_balanced, "heart_stroke_model.pkl")

print("Model saved successfully!")

Model saved successfully!


# Saving the Standard Scaler

The Logistic Regression model was trained on standardized features. Therefore, the fitted `StandardScaler` must also be saved so that new user inputs can be transformed in the same way before prediction.

In [34]:
# Save the Standard Scaler

joblib.dump(scaler, "scaler.pkl")

print("Scaler saved successfully!")

Scaler saved successfully!


# Creating Label Encoders

Separate Label Encoders are created for each categorical feature. These encoders preserve the mapping between category labels and numerical values, allowing the Streamlit application to encode user inputs consistently with the training data.

In [35]:
# Create Separate Label Encoders

from sklearn.preprocessing import LabelEncoder

gender_encoder = LabelEncoder()
education_encoder = LabelEncoder()
exercise_encoder = LabelEncoder()
agegroup_encoder = LabelEncoder()

gender_encoder.fit(heart_df["Gender"])
education_encoder.fit(heart_df["Education"])
exercise_encoder.fit(heart_df["Exercise"])
agegroup_encoder.fit(heart_df["AgeGroup"])

print("Encoders created successfully!")

Encoders created successfully!


# Saving Label Encoders

The fitted Label Encoders are saved using Joblib so that categorical user inputs can be converted into the same numerical values used during model training.

In [36]:
# Save Label Encoders

joblib.dump(gender_encoder, "gender_encoder.pkl")
joblib.dump(education_encoder, "education_encoder.pkl")
joblib.dump(exercise_encoder, "exercise_encoder.pkl")
joblib.dump(agegroup_encoder, "agegroup_encoder.pkl")

print("All encoders saved successfully!")

All encoders saved successfully!


# Verify Saved Files

The following files are generated and will be used by the Streamlit application during deployment.

In [37]:
import os

files = [
    "heart_stroke_model.pkl",
    "scaler.pkl",
    "gender_encoder.pkl",
    "education_encoder.pkl",
    "exercise_encoder.pkl",
    "agegroup_encoder.pkl"
]

for file in files:
    print(f"{file} :", os.path.exists(file))

heart_stroke_model.pkl : True
scaler.pkl : True
gender_encoder.pkl : True
education_encoder.pkl : True
exercise_encoder.pkl : True
agegroup_encoder.pkl : True


### Brainstorming

In [1]:
import pandas as pd

heart_df = pd.read_excel(r"D:\DS 8AM\health_disease_project\data\heart_disease_cleaned.xlsx")

In [2]:
print(sorted(heart_df["Education"].unique()))

['graduate', 'postgraduate', 'primaryschool', 'uneducated']


In [3]:
import joblib

education_encoder = joblib.load(
    r"D:\DS 8AM\health_disease_project\models\education_encoder.pkl"
)

print(education_encoder.classes_)

['graduate' 'postgraduate' 'primaryschool' 'uneducated']


In [4]:
import joblib

gender_encoder = joblib.load(r"D:\DS 8AM\health_disease_project\models\gender_encoder.pkl")
exercise_encoder = joblib.load(r"D:\DS 8AM\health_disease_project\models\exercise_encoder.pkl")
agegroup_encoder = joblib.load(r"D:\DS 8AM\health_disease_project\models\agegroup_encoder.pkl")

print("Gender:", gender_encoder.classes_)
print("Exercise:", exercise_encoder.classes_)
print("AgeGroup:", agegroup_encoder.classes_)

Gender: ['Female' 'Male']
Exercise: ['Daily' 'Monthly' 'Weekly']
AgeGroup: ['30-39' '40-49' '50-59' '60-70']


In [5]:
heart_df.columns.tolist()

['Gender',
 'Age',
 'Education',
 'currentSmoker',
 'cigsPerDay',
 'BPMeds',
 'prevalentStroke',
 'prevalentHyp',
 'Diabetes',
 'totChol',
 'sysBP',
 'diaBP',
 'BMI',
 'heartRate',
 'Glucose',
 'Exercise',
 'heartStroke',
 'AgeGroup']